hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java

```diff
@@ -496,12 +503,59 @@ SqlResult execute(ExplainStatementPlan plan) {
         List<Object> args = prepareArguments(plan.getParameterMetadata(), arguments);
+        InternalSerializationService serializationService = Util.getSerializationService(hazelcastInstance);
+        ExpressionEvalContext evalContext = new ExpressionEvalContextImpl(
+                args,
+                serializationService,
+                Util.getNodeEngine(hazelcastInstance));
+
         JobConfig jobConfig = new JobConfig()
                 .setArgument(SQL_ARGUMENTS_KEY_NAME, args)
                 .setArgument(KEY_SQL_QUERY_TEXT, plan.getQuery())
                 .setArgument(KEY_SQL_UNBOUNDED, plan.isStreaming())
                 .setTimeoutMillis(timeout);
 
+        final Set<Integer> partitions = new HashSet<>();
+        for (Entry<String, Map<String, Expression<?>>> candidate : plan.getPartitionStrategyCandidates().entrySet()) {
+            // TODO: null checks
+            final String mapName = candidate.getKey();
+            final Map<String, Expression<?>> keyExprs = candidate.getValue();
+            final MapContainer mapContainer = getMapContainer(hazelcastInstance.getMap(mapName));
+            final PartitioningStrategy strategy = mapContainer.getPartitioningStrategy() == null
+                    ? new DefaultPartitioningStrategy()
+                    : mapContainer.getPartitioningStrategy();
+
+            if (!(strategy instanceof AttributePartitioningStrategy)
+                    && !(strategy instanceof DefaultPartitioningStrategy)) {
+                // shouldn't happen
+                // TODO: maybe handle properly and bypass the candidates altogether if they're based on user's strategy?
+                throw new HazelcastException("Only Default and Attribute Partitioning Strategies are supported.");
+            }
+
+            final Object partitionKey;
+            if (keyExprs.size() > 1) {
+                assert strategy instanceof AttributePartitioningStrategy;
+                final AttributePartitioningStrategy attributeStrategy = (AttributePartitioningStrategy) strategy;
+                final Object[] attributes = Arrays.stream(attributeStrategy.getPartitioningAttributes())
+                        .map(keyExprs::get)
+                        // TODO assert that exprs are ConstantExpr or ParameterExpr
+                        .map(expr -> (Object) expr.eval(null, evalContext))
+                        .collect(Collectors.toList())
+                        .toArray(new Object[]{});
+                partitionKey = attributes;
+            } else {
+                assert candidate.getValue().size() == 1;
+                // TODO: different constant reference for __key
+                partitionKey = candidate.getValue().get(DEFAULT_UNIQUE_KEY);
+            }
+            // TODO: null checks to handle repartitioning
+            partitions.add(hazelcastInstance.getPartitionService().getPartition(partitionKey).getPartitionId());
```

Comment:
how about dynamic parameters determining partition? Will this be supported in this PR? If not, what is the idea? Will `KEY_REQUIRED_PARTITIONS` contain `Expression`s?

Reply:
OK, I got confused. Here we are executing query with specific argument values.